### ISSUE TITLE
Opened Tab/Accordion should have public URL

### ISSUE THREAD

**Issue by mkrecek234** [atk4/ui#2070]
Any directly reachable tab/accordion (ie. content without internal state/POST data like temporary modal) have VP/Callback link and:
- such link should be set as the current URL when clicked
- the link should be possible to be opened by using the link solely (of course with auth etc.)
- any menu items should use `a` tag with possibility to be opened like regular `<a href="...`, ie. with left/middle button, with context menu "Open in a New Tab/Window"...

---

Situation:
- You have a view which has tabs, and/or cruds and/or accordions.
- Sometimes it is helpful to send a URL of the exact screen the user is in to someone else. As the screen is the view script, but might also currently have an opened modal, or a selected tab or an opened accordion, it would be helpful that the URL reflects the current state of the view., conceptually:
- localhost/view.php?viewcrud1={useraction=edit, id=12345} for indicating that the user currently has opened an edit modal inside viewcrud1
- Sending then that URL will make sure, that on any browser the view is displayed in the same state as one sees it.

How to implement:
- the browsers URL-line dynamically will be adjusted via jQuery/DOM whenever a tab is selected, or a modal is opened or an accordion is opened; this makes sure that URL always reflects the right state as parameters to the current script.
- When rendering such a URL, the views like tab, grid/crud, accordion will respect the currently active configuration that is found in the URL.

**Comment by mkrecek234** [comment]
@mvorisek No, this is not completed. For example:
- Open and view with a crud, then select any line and open the edit modal.
- Then copy the URL in the browser window and open it in a new browser window.

Expected behaviour according to this feature request:
- The page is opened and it shows exactly the same view, including the opened edit modal

Current behavior:
- It opens the view without an opened modal.

**Comment by mvorisek** [comment]
If all params are passed by GET, the problem should be solved with/in #2068. Usecase with Modal, will require sending POST params in URL (as GET), do we wan to support that?

Last week I fixed https://github.com/atk4/ui/commit/a6400097792399feadce3358658670d8bf7bc341. It took me over a day to understand the problem... and I finally understand what we exactly do in atk4/ui multi step modal executor, we use local storage on the client... So again, this will need to be encoded in the URL. I am for removing the local storage, but I belive we should solve it gradually and with a budget we can reason by a better productivity, less maintanence, less end user errors, ... As you do not want to sponsor this issue completely right now, I close it again to keep the focus on important tasks.

**Comment by mkrecek234** [comment]
@mvorisek Can we please keep idea "issues" open - this should be a storage for ideas, even if they are not currently implemented.

**Comment by mvorisek** [comment]
If you want to implement this or contribute to this topic in any way, feel free to reopen it, but too broad ideas are not much helpful if only opened.

**Comment by mvorisek** [comment]
Michael, I narrowed your idea to something wanted and feasible and reopened it. Thank you.

**Comment by mkrecek234** [comment]
@mvorisek Thank you - I would like to make the idea more concrete that not only the tab has a public URL (on right-click) but also that the browser window‘s URL is adjusted on tab change, so whenever a user copies the current URL and forwards it, the recipient opens the same view state. This can be done with:

`const url = new URL(window.location.href);
url.searchParams.set('param1', 'val1');
url.searchParams.delete('param2');
window.history.replaceState(null, null, url); // or pushState`

**Comment by mvorisek** [comment]
yes, it is in the edited description already :)

**Comment by mvorisek** [comment]
This feature request is very tricky, as page can contains more than one tabs groups/accordions and can be nested - the URL will have to contain state for all of them.

**Comment by mkrecek234** [comment]
Yes, correct, but in a lot of UIs for usability reasons you would not nest too many tabs or accordions. Also a simple solution would be to make the accordion or tab group configurable if it should update the URL and add its state to it or not. In this case the developer could set which accirdion or tab can be pre-set by a URL parameter.

**Comment by mvorisek** [comment]
For single tab/traverse path - related issues: https://github.com/atk4/ui/issues/451 and https://github.com/atk4/mastercrud/issues/25
